src/plugins/expressions/common/execution/execution.ts

```diff
@@ -190,23 +197,55 @@ export class Execution<
       }
 
       const { function: fnName, arguments: fnArgs } = link;
-      const fnDef = getByAlias(this.state.get().functions, fnName);
+      const fn = getByAlias(this.state.get().functions, fnName);
 
-      if (!fnDef) {
+      if (!fn) {
         return createError({ message: `Function ${fnName} could not be found.` });
       }
 
+      let args: Record<string, ExpressionValue> = {};
+      let timeStart: number | undefined;
+
       try {
-        // Resolve arguments before passing to function
-        // resolveArgs returns an object because the arguments themselves might
-        // actually have a 'then' function which would be treated as a promise
-        const { resolvedArgs } = await this.resolveArgs(fnDef, input, fnArgs);
-        const output = await this.invokeFunction(fnDef, input, resolvedArgs);
+        // `resolveArgs` returns an object because the arguments themselves might
+        // actually have a `then` function which would be treated as a `Promise`.
+        const { resolvedArgs } = await this.resolveArgs(fn, input, fnArgs);
+        args = resolvedArgs;
+        timeStart = this.params.debug ? performance.now() : 0;
+        const output = await this.invokeFunction(fn, input, resolvedArgs);
+
+        if (this.params.debug) {
+          const timeEnd: number = performance.now();
+          (link as ExpressionAstFunction).debug = {
```

Comment:
rather than adding it to the AST should we store this information in inspector adapters ? (create specific adapter for this)

Reply:
We can still create an inspector adapter using this approach. If we don't store debug information on AST, where will we store it? We would need to create another AST or "execution chain", [like described in the issue](https://github.com/elastic/kibana/issues/56746):

> - [x] Debug information could be collected on:
>   - ~~An "execution chain"&mdash;a linked list of execution chain is created with debug information about every function as they execute. Note: this chain will contain information only until the first function that *throws* (because once one function throws expression executor stops execution)~~.
>   - [x] Store debug information on the AST used to execute the function.
>   - ~~Create a new AST as expression executes~~.

I chose to store the info on AST, because it is the simplest, it basically just adds two `IfStatements` to the codebase.

Reply:
in that case the inspector adapter would be the execution chain mentioned in the issue.
right now we would need to copy the ast in the inspector adapter, which is also fine.